mongocrypt/src/ctx.rs

```diff
@@ -0,0 +1,687 @@
+    }
+}
+
+/// Manages the state machine for encryption or decryption.
+impl Ctx {
+    /// Get the current state of a context.
+    pub fn state(&self) -> Result<State> {
+        let s = unsafe { sys::mongocrypt_ctx_state(*self.inner.borrow()) };
+        if s == sys::mongocrypt_ctx_state_t_MONGOCRYPT_CTX_ERROR {
+            return Err(self.status().as_error());
+        }
+        Ok(State::from_native(s))
+    }
+
+    /// Get BSON necessary to run the mongo operation when in `State::NeedMongo*` states.
+    ///
+    /// The returned value:
+    /// * for `State::NeedMongoCollinfo` it is a listCollections filter.
+    /// * for `State::NeedMongoKeys` it is a find filter.
+    /// * for `State::NeedMongoMarkings` it is a command to send to mongocryptd.
+    pub fn mongo_op(&self) -> Result<&RawDocument> {
+        // Safety: `mongocrypt_ctx_mongo_op` updates the passed-in `Binary` to point to a chunk of
+        // BSON with the same lifetime as the underlying `Ctx`.  The `Binary` itself does not own
+        // the memory, and gets cleaned up at the end of the unsafe block.  Lifetime inference on
+        // the return type binds `op_bytes` to the same lifetime as `&self`, which is the correct
+        // one.
+        let op_bytes = unsafe {
+            let bin = Binary::new();
+            if !sys::mongocrypt_ctx_mongo_op(*self.inner.borrow(), *bin.native()) {
+                return Err(self.status().as_error());
+            }
+            bin.bytes()?
+        };
+        rawdoc(op_bytes)
+    }
+
+    /// Feed a BSON reply or result when this context is in
+    /// `State::NeedMongo*` states. This may be called multiple times
+    /// depending on the operation.
+    ///
+    /// `reply` is a BSON document result being fed back for this operation.
+    /// - For `State::NeedMongoCollinfo` it is a doc from a listCollections
+    /// cursor. (Note, if listCollections returned no result, do not call this
+    /// function.)
+    /// - For `State::NeedMongoKeys` it is a doc from a find cursor.
+    ///   (Note, if find returned no results, do not call this function.)
+    /// - For `State::NeedMongoMarkings` it is a reply from mongocryptd.
+    pub fn mongo_feed(&mut self, reply: &RawDocument) -> Result<()> {
+        let bin = BinaryRef::new(reply.as_bytes());
+        unsafe {
+            if !sys::mongocrypt_ctx_mongo_feed(*self.inner.borrow(), *bin.native()) {
+                return Err(self.status().as_error());
+            }
+        }
+        Ok(())
+    }
+
+    /// Call when done feeding the reply (or replies) back to the context.
+    pub fn mongo_done(&mut self) -> Result<()> {
+        unsafe {
+            if !sys::mongocrypt_ctx_mongo_done(*self.inner.borrow()) {
+                return Err(self.status().as_error());
+            }
+        }
+        Ok(())
+    }
+
+    pub fn kms_scope(&mut self) -> KmsScope {
+        KmsScope { ctx: self }
+    }
+
+    /// all in response to the `State::NeedKmsCredentials` state
```

Comment:
Did the beginning of the docstring get cut off here?

Reply:
Oops, fixed.